### PR TITLE
onboarding lab implementation

### PR DESCRIPTION
**Pull request by dagitses**

### CODE REVIEW

File: aten/src/ATen/native/Attn.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include <tuple>
+
+#include <ATen/Tensor.h>
+
+namespace at { namespace native {
+
+auto attn(Tensor const& q, Tensor const& k, Tensor const& v) -> std::tuple<Tensor, Tensor> {
+  Tensor attn = q.matmul(k.t()).tanh();
```
Comment (albanD) [inline aten/src/ATen/native/Attn.cpp:8]: It would be nice to have some error checking here, to make sure that all the inputs have the right number of dimensions at least.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -717,6 +717,8 @@
 - func: atleast_3d.Sequence(Tensor[] tensors) -> Tensor[]
   variants: function
 
+- func: attn(Tensor q, Tensor k, Tensor v) -> (Tensor, Tensor)
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:720]: Note that if you actually put some names for the output Tensors, it will automatically become a named tuple in python. Which makes it much nicer!

**Review by albanD — COMMENTED** [review]
It would be nice to have at least a simple call to the new function in the first commit. Just to make sure it works :p

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -1390,6 +1390,25 @@ def sample_inputs_addcmul_addcdiv(op_info, device, dtype, requires_grad, **kwarg
...
+    # Three inputs: q, k, v that are all 2-D tensors (m × n).
+    # qₘ = kₘ = vₘ = m
+    # qₙ = kₙ = n
+    # vₙ is unconstrained
+    #
+    # Thus there are three shape variables that can vary
+    # independently: {m, n, vₙ}. The Cartesian cube of an integral set,
+    # sizes here, generates test cases for (m, n, vₙ).
+    sizes = torch.tensor([1, S, M, L])
+    test_cases = torch.cartesian_prod(sizes, sizes, sizes)
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:1403]: +1 point for using torch functions only !
If you want to do that with python lists, you can use `itertools.product()` as well

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -4949,6 +4968,12 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                SkipInfo('TestCommon', 'test_variant_consistency_eager'),
            ),
            sample_inputs_func=sample_inputs_addbmm),
+    OpInfo('attn',
+           dtypes=floating_types_and(torch.bfloat16),
+           backward_dtypes=floating_types_and(torch.bfloat16),
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:4973]: You don't need to specify this one if it matches the one above right?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -4949,6 +4968,12 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+           supports_forward_ad=False,
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:4974]: This should be the default.

File: tools/autograd/derivatives.yaml [hunk 6]
```diff
@@ -302,6 +302,10 @@
 - name: atan2(Tensor self, Tensor other) -> Tensor
   self, other: atan2_backward(grad, self, other, grad_input_mask)
 
+- name: attn(Tensor q, Tensor k, Tensor v) -> (Tensor, Tensor)
+  q, k: attn_backward(grads[0], grads[1], q, k, v, result1)
+  v: 'grads[0].defined() ? result1.t().matmul(grads[0]) : zeros_like(v)'
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:307]: For gradients, an undefined Tensor is the same as a Tensor full of zeros. So the second case here can be `Tensor()` and that will be much more efficient.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -1962,6 +1962,27 @@ std::tuple<Tensor, Tensor> atan2_backward(const Tensor& grad, const Tensor& self
...
 
+std::tuple<Tensor, Tensor> attn_backward(Tensor grad_output, Tensor grad_attn,
+                                         const Tensor& q, const Tensor& k, const Tensor& v,
+                                         const Tensor& attn) {
+  // u3 in the forward pass is attn.matmul(v), aka output. We use a
+  // different name because grad_output is already present as the
+  // gradient of the output result.
+  Tensor grad_u3 = grad_output.defined()
+    ? grad_output.matmul(v.t())
+    : at::zeros({attn.size(0), v.size(0)});  // match dimensions above
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1973]: Could you do better than materializing a Tensor full of zeros here? (with the same idea that undefined Tensor is the same as full of zeros for gradients).

File: aten/src/ATen/native/Attn.cpp (hunk 1 above)
Comment (dagitses) [reply aten/src/ATen/native/Attn.cpp:27]: done.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (dagitses) [reply torch/testing/_internal/common_methods_invocations.py:1403]: does the +1 mean that this was a good way to do it? my instinct would be to do it in pure python if i'm not benefiting from pytorch stuff, but is that the wrong instinct?

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (dagitses) [reply torch/testing/_internal/common_methods_invocations.py]: correct! they used to differ, hence the origin of this.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (dagitses) [reply aten/src/ATen/native/native_functions.yaml]: neat!

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (dagitses) [reply torch/testing/_internal/common_methods_invocations.py]: correct, removed.

File: tools/autograd/derivatives.yaml (hunk 6 above)
Comment (dagitses) [reply tools/autograd/derivatives.yaml]: easy enough.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 7 above)
Comment (dagitses) [reply torch/csrc/autograd/FunctionsManual.cpp]: i can. i went a little overboard here with doing things in-place where allowed. I'm not sure why I can't make the final transform in-place. Any ideas?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:1403]: It's a style point :D
But yes I think it would be nicer to read in pure python.

File: test/test_attn.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import torch
+
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+def attn(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> torch.Tensor:
+    attn = q.matmul(k.t()).tanh()
+    return attn.matmul(v), attn
+
+class TestAttn(TestCase):
```
Comment (albanD) [inline test/test_attn.py:9]: Note that for this file to run in CI, you need to add it to the list in `test/run_test.py`

File: aten/src/ATen/native/Attn.cpp [hunk 9]
```diff
@@ -0,0 +1,33 @@
+#include <tuple>
+
+#include <ATen/Tensor.h>
+
+namespace at { namespace native {
+
+namespace {
+
+auto is_matrix(Tensor const& t) -> bool {
+  return t.sizes().size() == 2;
```
Comment (albanD) [inline aten/src/ATen/native/Attn.cpp:10]: nit: you also have `.ndim()` here

File: test/test_attn.py (hunk 8 above)
Comment (dagitses) [reply test/test_attn.py:9]: done.

File: aten/src/ATen/native/Attn.cpp (hunk 9 above)
Comment (dagitses) [reply aten/src/ATen/native/Attn.cpp]: looks like i need .dim() in c++, ndim is only an alias available in python.

I find functions like this as members of TensorImpl?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (dagitses) [reply torch/testing/_internal/common_methods_invocations.py]: done.

File: aten/src/ATen/native/Attn.cpp (hunk 9 above)
Comment (albanD) [reply aten/src/ATen/native/Attn.cpp]: Yes

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -4992,6 +5012,11 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                SkipInfo('TestCommon', 'test_variant_consistency_eager'),
            ),
            sample_inputs_func=sample_inputs_addbmm),
+    OpInfo('attn',
+           dtypes=floating_types_and(torch.bfloat16),
+           dtypesIfCUDA=floating_types_and(torch.float16, *[torch.bfloat16] if CUDA11OrLater else []),
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:5017]: Nice fix.
We should fix that for future bootcampers!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 11]
```diff
@@ -1961,6 +1961,41 @@ std::tuple<Tensor, Tensor> atan2_backward(const Tensor& grad, const Tensor& self
...
+    grad_u3 = grad_attn;
+    in_place = false;
+  } else {
+    return std::make_tuple(Tensor(), Tensor());
+  }
+
+  TORCH_INTERNAL_ASSERT(grad_u3.defined());
+
+  // Conditionally point-wise multiplies tensors in-place.
+  auto mul = [](bool in_place, Tensor a, const Tensor& b) {
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1988]: I guess this is not needed anymore as you use it only once?

**Review by albanD — APPROVED** [review]
LGTM!

Just a nit but the code looks good otherwise!

File: torch/testing/_internal/common_methods_invocations.py (hunk 10 above)
Comment (dagitses) [reply torch/testing/_internal/common_methods_invocations.py:5017]: out of curiosity, what needs to be fixed here? i cargo-culted this line from other ops.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 11 above)
Comment (dagitses) [reply torch/csrc/autograd/FunctionsManual.cpp]: Done. FWIW, I think it reads better with the function, but it's subjective and marginal either way.

**Comment by albanD** [comment]
Looks like all CI is green (appart two flaky ones).
If you're satisfied with this, we can close a consider this done! :)
